src/sparse_blas/backends/mkl_common/mkl_helper.hpp

```diff
@@ -1,56 +1,102 @@
+#define _ONEMKL_SRC_SPARSE_BLAS_BACKENDS_MKL_COMMON_MKL_HELPER_HPP_
 
-// Includes are set up so that oneapi::mkl::sparse namespace refers to the MKLCPU and MKLGPU backends namespace (oneMKL product)
-// in this file.
-// oneapi::mkl::sparse::detail namespace refers to the oneMKL interface namespace.
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
 
+#include "oneapi/mkl/exceptions.hpp"
 #include "oneapi/mkl/sparse_blas/detail/helper_types.hpp"
 
+#include "sparse_blas/enum_data_types.hpp"
+#include "sparse_blas/macros.hpp"
+
 namespace oneapi::mkl::sparse::detail {
 
-inline auto get_handle(detail::matrix_handle **handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t *>(handle);
+/// Return whether a pointer is accessible on the host
+template <typename T>
+inline bool is_ptr_accessible_on_host(sycl::queue &queue, const T *host_or_device_ptr) {
+    auto alloc_type = sycl::get_pointer_type(host_or_device_ptr, queue.get_context());
+    // Note sycl::usm::alloc::host may not be accessible on the host according to SYCL specification.
+    // sycl::usm::alloc::unknown is returned if the pointer is not a USM allocation which is assumed to be a normal host pointer.
+    return alloc_type == sycl::usm::alloc::shared || alloc_type == sycl::usm::alloc::unknown;
+}
+
+/// Throw an exception if the scalar is not accessible in the host
+template <typename T>
+void check_ptr_is_host_accessible(const std::string &function_name, const std::string &scalar_name,
+                                  sycl::queue &queue, const T *host_or_device_ptr) {
+    if (!is_ptr_accessible_on_host(queue, host_or_device_ptr)) {
+        throw mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "Scalar " + scalar_name + " must be accessible on the host for buffer functions.");
+    }
 }
 
-inline auto get_handle(detail::matrix_handle *handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t>(handle);
+/// Return a scalar on the host from a pointer to host or device memory
+/// Used for USM functions
+template <typename T>
+inline T get_scalar(sycl::queue &queue, const T *host_or_device_ptr) {
+    if (is_ptr_accessible_on_host(queue, host_or_device_ptr)) {
+        return *host_or_device_ptr;
+    }
+    T scalar;
+    auto event = queue.copy(host_or_device_ptr, &scalar, 1);
+    event.wait_and_throw();
+    return scalar;
 }
 
+/// Merge multiple event dependencies into one
+inline sycl::event collapse_dependencies(sycl::queue &queue,
+                                         const std::vector<sycl::event> &dependencies) {
+    if (dependencies.empty()) {
+        return {};
+    }
+    else if (dependencies.size() == 1) {
+        return dependencies[0];
+    }
+
+    return queue.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(dependencies);
+        cgh.host_task([=]() {});
+    });
+}
```

Comment:
I guess it's unfortunate we can't use [`ext_oneapi_submit_barrier()`](https://intel.github.io/llvm-docs/doxygen/classsycl_1_1__V1_1_1queue.html#a9c7a0dc5dd031073e00afd7b5e020761) API yet because it's not part of the SYCL specification (yet).

Reply:
Sure, looking at this [documentation](https://github.com/intel/llvm/blob/417cc1b18538c4f9c028018bc77093c2a6c8cba9/sycl/doc/extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc#overview) this extension is to simplify the user's code. I don't expect this would make much of a difference to oneMKL Interface.